src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -143,4 +171,34 @@ public AsyncJsonTokenParser(
     return Futures.transformAsync(futureVerifiers, checkNullFunction, executor);
   }
 
+  /**
+   * Converts exceptions (when applicable) to {@link InvalidJsonTokenException}
+   * for better exception handling in the asynchronous parser.
+   *
+   * @return a failed {@link ListenableFuture}
+   */
+  private ListenableFuture getFailedFuture(Exception e) {
+    Throwable cause = e.getCause();
+    if (cause != null && InvalidJsonTokenException.class.isInstance(cause)) {
+      InvalidJsonTokenException invalidJsonTokenException = (InvalidJsonTokenException) cause;
+      if (invalidJsonTokenException.getErrorCode().equals(ErrorCode.ILLEGAL_STATE)) {
+        return Futures.immediateFailedFuture(e);
+      }
+      return Futures.immediateFailedFuture(
+          new InvalidJsonTokenException(invalidJsonTokenException.getErrorCode(), e));
+    }
+
+    if (JsonParseException.class.isInstance(e)) {
+      return Futures.immediateFailedFuture(
+          new InvalidJsonTokenException(ErrorCode.MALFORMED_TOKEN_STRING, e));
+    }
+
+    if (RuntimeException.class.isInstance(e)) {
+      return Futures.immediateFailedFuture(e);
+    }
+
+    return Futures.immediateFailedFuture(
+        new InvalidJsonTokenException(ErrorCode.UNKNOWN, e));
```

Comment:
When is this used besides when e is SignatureException? I'm a little hesitant to replace all non-runtime exceptions with InvalidJsonTokenException and would rather only replace the exception for known/expected exceptions and leave the original exception as-is otherwise (the error code UNKNOWN is used because we don't know the exact reason why SignatureException was thrown since it was thrown by one of the checkers, but I don't think we should blindly replace all exceptions with UNKNOWN).

Reply:
I don't think we know of checked exceptions other than SignatureException in the classes that have been modified. I will change it to only explicitly check for SignatureException just in case.